pkg/fanal/artifact/image/image.go

```diff
@@ -277,9 +277,13 @@ func (a Artifact) inspectLayer(ctx context.Context, layerInfo LayerInfo, disable
 	}
 	defer composite.Cleanup()
 
+	if err := artifact.AddConfigFilesToFS(composite, a.artifactOption); err != nil {
```

Comment:
IMO, we should not call a misconfiguration-specific method here because we must call it in all artifacts. For example, this PR is not calling it in VM artifacts. It can lead to a bug.

What if putting files in misconf/scanner.go? Then, each artifact doesn't have to care about IaC config files.

```
diff --git a/pkg/misconf/scanner.go b/pkg/misconf/scanner.go
index 72f97f968..8f76b739c 100644
--- a/pkg/misconf/scanner.go
+++ b/pkg/misconf/scanner.go
@@ -71,6 +71,7 @@ type Scanner struct {
        fileType       detection.FileType
        scanner        scanners.FSScanner
        hasFilePattern bool
+       configFiles    []string
 }

 func NewAzureARMScanner(filePatterns []string, opt ScannerOption) (*Scanner, error) {
@@ -108,6 +109,7 @@ func newScanner(t detection.FileType, filePatterns []string, opt ScannerOption)
        }

        var scanner scanners.FSScanner
+       var configFiles []string
        switch t {
        case detection.FileTypeAzureARM:
                scanner = arm.New(opts...)
@@ -117,10 +119,12 @@ func newScanner(t detection.FileType, filePatterns []string, opt ScannerOption)
                scanner = dfscanner.NewScanner(opts...)
        case detection.FileTypeHelm:
                scanner = helm.New(opts...)
+               configFiles = append(opt.HelmValueFiles, opt.HelmValueFiles...)
        case detection.FileTypeKubernetes:
                scanner = k8sscanner.NewScanner(opts...)
        case detection.FileTypeTerraform:
                scanner = tfscanner.New(opts...)
+               configFiles = opt.TerraformTFVars
        case detection.FileTypeTerraformPlan:
                scanner = tfpscanner.New(opts...)
        }
@@ -129,6 +133,7 @@ func newScanner(t detection.FileType, filePatterns []string, opt ScannerOption)
                fileType:       t,
                scanner:        scanner,
                hasFilePattern: hasFilePattern(t, filePatterns),
+               configFiles:    configFiles,
        }, nil
 }

@@ -141,6 +146,9 @@ func (s *Scanner) Scan(ctx context.Context, fsys fs.FS) ([]types.Misconfiguratio
                return nil, nil
        }

+       // Add config files to newfs here
+       // addConfigFilesToFS(newfs, s.configFiles)
+
        log.Logger.Debugf("Scanning %s files for misconfigurations...", s.scanner.Name())
        results, err := s.scanner.ScanFS(ctx, newfs, ".")
        if err != nil {
@@ -196,6 +204,7 @@ func (s *Scanner) filterFS(fsys fs.FS) (fs.FS, error) {
        if !foundRelevantFile {
                return nil, nil
        }
+
        return newfs, nil
 }
```

Reply:
But it adds files to the existing filesystem. We may want to create a dedicated filesystem for config files and pass it to scanners.

Reply:
So what do you propose? Should we merge this and create a new issue to track the improvement for a dedicated filesystem for config files?

Reply:
Right. That is in my mind.

Reply:
Cool! I created https://github.com/aquasecurity/trivy/issues/5280 to track that change. Can we merge this PR for now?

Reply:
I'd apply this kind of change regardless of #5280.
https://github.com/aquasecurity/trivy/pull/5218#discussion_r1337664734